File: android/app/src/main/java/com/targetzone/library/ui/student/StudentViewModel.kt
package com.targetzone.library.ui.student

import android.app.DownloadManager
import android.content.Context
import android.net.Uri
import android.os.Environment
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.targetzone.library.data.TokenManager
import com.targetzone.library.data.api.BASE_URL
import com.targetzone.library.data.model.*
import com.targetzone.library.data.repository.GalleryRepository
import com.targetzone.library.data.repository.MembershipRepository
import com.targetzone.library.data.repository.SeatRepository
import com.targetzone.library.data.repository.UserRepository
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asSharedFlow
import kotlinx.coroutines.launch
import java.io.File

class StudentViewModel(
    private val membershipRepo: MembershipRepository = MembershipRepository(),
    private val seatRepo: SeatRepository = SeatRepository(),
    private val userRepo: UserRepository = UserRepository(),
    private val galleryRepo: GalleryRepository = GalleryRepository(),
    private val tokenManager: TokenManager? = null
) : ViewModel() {

    val membership    = MutableStateFlow<Membership?>(null)
    val plans         = MutableStateFlow<List<Plan>>(emptyList())
    val seats         = MutableStateFlow<List<Seat>>(emptyList())
    val selectedSeat  = MutableStateFlow<Seat?>(null)
    val profile       = MutableStateFlow<User?>(null)
    val myFeedback    = MutableStateFlow<List<FeedbackItem>>(emptyList())
    val membershipHistory = MutableStateFlow<List<Membership>>(emptyList())
    val galleryPhotos = MutableStateFlow<List<GalleryPhoto>>(emptyList())

    val isLoading     = MutableStateFlow(false)
    val error         = MutableStateFlow<String?>(null)

    // Payment flow: emit PaymentOrder when ready, MainActivity triggers Razorpay
    private val _paymentOrder = MutableSharedFlow<PaymentOrder>()
    val paymentOrder = _paymentOrder.asSharedFlow()

    val bookingSuccess = MutableStateFlow(false)

    fun loadDashboard() = viewModelScope.launch {
        membershipRepo.getMyMembership()
            .onSuccess { membership.value = it }
    }

    fun loadPlans() = viewModelScope.launch {
        membershipRepo.getPlans().onSuccess { plans.value = it }
    }

    fun loadSeats(shift: String) = viewModelScope.launch {
        isLoading.value = true
        seatRepo.getAvailability(shift)
            .onSuccess { seats.value = it }
            .onFailure { error.value = it.message }
        isLoading.value = false
    }

    fun selectSeat(seat: Seat?) { selectedSeat.value = seat }

    fun startPayment(planId: String, seatNumber: String, shift: String) = viewModelScope.launch {
        isLoading.value = true
        error.value = null
        membershipRepo.createOrder(planId, seatNumber, shift)
            .onSuccess { order ->
                isLoading.value = false
                if (order.orderId.startsWith("dev_")) {
                    // Dev mode: skip Razorpay, verify directly
                    verifyPayment(order.orderId, "dev_pay_${System.currentTimeMillis()}", "dev_sig", order.membershipId)
                } else {
                    _paymentOrder.emit(order)
                }
            }
            .onFailure { isLoading.value = false; error.value = it.message }
    }

    fun verifyPayment(gatewayOrderId: String, paymentId: String, signature: String, membershipId: String) = viewModelScope.launch {
        isLoading.value = true
        membershipRepo.verifyPayment(VerifyPaymentRequest(gatewayOrderId, paymentId, signature, membershipId))
            .onSuccess { m ->
                membership.value = m
                isLoading.value = false
                bookingSuccess.value = true
            }
            .onFailure { isLoading.value = false; error.value = it.message }
    }

    fun loadProfile() = viewModelScope.launch {
        userRepo.getProfile().onSuccess { profile.value = it }
    }

    fun updateProfile(req: UpdateProfileRequest) = viewModelScope.launch {
        isLoading.value = true
        userRepo.updateProfile(req)
            .onSuccess { profile.value = it; isLoading.value = false }
            .onFailure { error.value = it.message; isLoading.value = false }
    }

    fun uploadPhoto(file: File) = viewModelScope.launch {
        isLoading.value = true
        userRepo.uploadPhoto(file)
            .onSuccess { loadProfile(); isLoading.value = false }
            .onFailure { error.value = it.message; isLoading.value = false }
    }

    fun uploadAadhaar(file: File) = viewModelScope.launch {
        isLoading.value = true
        userRepo.uploadAadhaar(file)
            .onSuccess { loadProfile(); isLoading.value = false }
            .onFailure { error.value = it.message; isLoading.value = false }
    }

    fun downloadIdCard(context: Context) = viewModelScope.launch {
        val token = tokenManager?.getTokenBlocking() ?: return@launch
        val dm = context.getSystemService(Context.DOWNLOAD_SERVICE) as DownloadManager
        val request = DownloadManager.Request(Uri.parse("${BASE_URL}memberships/my/id-card"))
            .addRequestHeader("Authorization", "Bearer $token")
            .setDestinationInExternalPublicDir(Environment.DIRECTORY_DOWNLOADS, "library-id-card.pdf")
            .setNotificationVisibility(DownloadManager.Request.VISIBILITY_VISIBLE_NOTIFY_COMPLETED)
            .setMimeType("application/pdf")
        dm.enqueue(request)
    }

    fun loadFeedback() = viewModelScope.launch {
        membershipRepo.getMyFeedback().onSuccess { myFeedback.value = it }
    }

    fun submitFeedback(type: String, subject: String, description: String, onDone: () -> Unit) = viewModelScope.launch {
        isLoading.value = true
        membershipRepo.submitFeedback(type, subject, description)
            .onSuccess { isLoading.value = false; loadFeedback(); onDone() }
            .onFailure { error.value = it.message; isLoading.value = false }
    }

    fun loadMembershipHistory() = viewModelScope.launch {
        membershipRepo.getMembershipHistory().onSuccess { membershipHistory.value = it }
    }

    fun loadGallery() = viewModelScope.launch {
        galleryRepo.getAll()
            .onSuccess { galleryPhotos.value = it }
            .onFailure { error.value = it.message }
    }

    fun resetBooking() { selectedSeat.value = null; bookingSuccess.value = false; error.value = null }
    fun clearError() { error.value = null }
}

File: ios/LibrarySystem/LibrarySystem/Features/Student/StudentViewModel.swift
import Foundation
import Combine

@MainActor
final class StudentViewModel: ObservableObject {
    // Dashboard
    @Published var membership: Membership?
    @Published var queuedMembership: Membership?
    @Published var membershipHistory: [Membership] = []

    // Plans & Booking
    @Published var plans: [Plan] = []
    @Published var selectedPlan: Plan?
    @Published var selectedShift = "MORNING"
    @Published var seats: [Seat] = []
    @Published var selectedSeat: String?
    @Published var pendingOrder: PaymentOrder?

    // Renew/queue next plan (kept separate from pendingOrder above so the
    // Booking tab's payment sheet never reacts to an order created from the
    // Membership tab's renew flow, since both tabs stay alive under TabView)
    @Published var queueOrder: PaymentOrder?
    @Published var queuePaying = false

    // Pay off grace-period dues
    @Published var duesOrder: PaymentOrder?
    @Published var payingDues = false

    // Profile
    @Published var profile: User?
    @Published var photoUrl: String?
    @Published var aadhaarUrl: String?

    // Feedback
    @Published var myFeedback: [FeedbackItem] = []

    // Payment history
    @Published var myPayments: [StudentPayment] = []

    // Gallery
    @Published var galleryPhotos: [GalleryPhoto] = []

    // Contact Admin
    @Published var adminContact: AdminContact?
    @Published var callAdminSent = false

    // UI State
    @Published var isLoading = false
    @Published var error: String?
    @Published var successMsg: String?
    @Published var idCardData: Data?
    @Published var showIdCardShare = false

    private let membershipRepo = MembershipRepository.shared
    private let userRepo       = UserRepository.shared
    private let seatRepo       = SeatRepository.shared

    // MARK: - Dashboard

    func loadDashboard() {
        isLoading = true; error = nil
        Task {
            do { membership = try await membershipRepo.getMyMembership() }
            catch { /* no active membership is a normal 404 */ }
            queuedMembership = await membershipRepo.getQueuedMembership()
            isLoading = false
        }
    }

    func loadPlans() {
        Task {
            do { plans = try await membershipRepo.getPlans() }
            catch { self.error = error.localizedDescription }
        }
    }

    // MARK: - Seats

    func loadSeats(shift: String, date: String? = nil) {
        isLoading = true
        Task {
            do { seats = try await seatRepo.getAvailability(shift: shift, date: date) }
            catch { self.error = error.localizedDescription }
            isLoading = false
        }
    }

    func selectSeat(_ seatNumber: String) {
        selectedSeat = seatNumber
    }

    // MARK: - Payment

    func startPayment(planId: String, seatNumber: String, shift: String) {
        isLoading = true; error = nil
        Task {
            do {
                pendingOrder = try await membershipRepo.createOrder(
                    planId: planId, seatNumber: seatNumber, shift: shift)
            } catch { self.error = error.localizedDescription }
            isLoading = false
        }
    }

    func verifyPayment(gatewayOrderId: String, gatewayPaymentId: String,
                       signature: String, membershipId: String) {
        isLoading = true
        Task {
            do {
                membership = try await membershipRepo.verifyPayment(
                    gatewayOrderId: gatewayOrderId,
                    gatewayPaymentId: gatewayPaymentId,
                    signature: signature,
                    membershipId: membershipId
                )
                resetBooking()
                successMsg = "Payment successful!"
            } catch { self.error = error.localizedDescription }
            isLoading = false
        }
    }

    func devVerifyPayment(membershipId: String) {
        verifyPayment(gatewayOrderId: pendingOrder?.orderId ?? "",
                      gatewayPaymentId: "dev_pay_\(membershipId)",
                      signature: "dev_sig",
                      membershipId: membershipId)
    }

    func resetBooking() {
        selectedPlan = nil
        selectedShift = "MORNING"
        selectedSeat = nil
        seats = []
        pendingOrder = nil
    }

    // MARK: - Renew / Queue Next Plan
    // Reuses the same create-order/verify endpoints as a fresh booking — the
    // backend auto-detects an existing active membership and queues the new
    // one, inheriting seat and shift, instead of activating it immediately.

    func startQueuePayment(planId: String) {
        queuePaying = true; error = nil
        Task {
            do { queueOrder = try await membershipRepo.createOrder(planId: planId, seatNumber: "", shift: "") }
            catch { self.error = error.localizedDescription }
            queuePaying = false
        }
    }

    func verifyQueuePayment(gatewayOrderId: String, gatewayPaymentId: String,
                            signature: String, membershipId: String) {
        queuePaying = true
        Task {
            do {
                let queued = try await membershipRepo.verifyPayment(
                    gatewayOrderId: gatewayOrderId, gatewayPaymentId: gatewayPaymentId,
                    signature: signature, membershipId: membershipId)
                if !queued.seatNumber.isEmpty {
                    do {
                        try await seatRepo.bookSeat(seatNumber: queued.seatNumber, membershipId: queued.id,
                                                    shift: queued.shift, startDate: queued.startDate,
                                                    endDate: queued.endDate)
                    } catch {
                        self.error = "Payment succeeded, but seat reservation failed — please contact support to finalize your seat."
                    }
                }
                queuedMembership = queued
                queueOrder = nil
                successMsg = "Plan queued! It will activate when your current plan expires."
            } catch { self.error = error.localizedDescription }
            queuePaying = false
        }
    }

    func devVerifyQueuePayment(membershipId: String) {
        verifyQueuePayment(gatewayOrderId: queueOrder?.orderId ?? "",
                           gatewayPaymentId: "dev_pay_\(membershipId)",
                           signature: "dev_sig", membershipId: membershipId)
    }

    func resetQueuePayment() {
        queueOrder = nil
    }

    // MARK: - Pay Grace Dues

    func startDuesPayment() {
        payingDues = true; error = nil
        Task {
            do { duesOrder = try await membershipRepo.createDuesOrder() }
            catch { self.error = error.localizedDescription }
            payingDues = false
        }
    }

    func verifyDuesPayment(gatewayOrderId: String, gatewayPaymentId: String,
                           signature: String, membershipId: String) {
        payingDues = true
        Task {
            do {
                membership = try await membershipRepo.verifyDuesPayment(
                    gatewayOrderId: gatewayOrderId, gatewayPaymentId: gatewayPaymentId,
                    signature: signature, membershipId: membershipId)
                duesOrder = nil
                successMsg = "Dues cleared — your membership is active again!"
            } catch { self.error = error.localizedDescription }
            payingDues = false
        }
    }

    func devVerifyDuesPayment(membershipId: String) {
        verifyDuesPayment(gatewayOrderId: duesOrder?.orderId ?? "",
                          gatewayPaymentId: "dev_pay_\(membershipId)",
                          signature: "dev_sig", membershipId: membershipId)
    }

    func resetDuesPayment() {
        duesOrder = nil
    }

    // MARK: - Profile

    func loadProfile() {
        Task {
            do {
                profile = try await userRepo.getProfile()
                photoUrl = profile?.photoUrl
                aadhaarUrl = profile?.aadhaarUrl
            } catch { self.error = error.localizedDescription }
        }
    }

    func updateProfile(name: String, fatherName: String?, address: String?,
                       gender: String?, dateOfBirth: String?, email: String?) {
        isLoading = true
        Task {
            do {
                profile = try await userRepo.updateProfile(
                    name: name, fatherName: fatherName, address: address,
                    gender: gender, dateOfBirth: dateOfBirth, email: email)
                successMsg = "Profile updated"
            } catch { self.error = error.localizedDescription }
            isLoading = false
        }
    }

    func uploadPhoto(data: Data) {
        Task {
            do {
                let url = try await userRepo.uploadPhoto(data: data)
                photoUrl = url
                successMsg = "Photo updated"
            } catch { self.error = error.localizedDescription }
        }
    }

    func uploadAadhaar(data: Data) {
        Task {
            do {
                let url = try await userRepo.uploadAadhaar(data: data)
                aadhaarUrl = url
                successMsg = "Aadhaar uploaded"
            } catch { self.error = error.localizedDescription }
        }
    }

    func downloadIdCard() {
        Task {
            do {
                idCardData = try await membershipRepo.downloadIdCard()
                showIdCardShare = true
            } catch { self.error = error.localizedDescription }
        }
    }

    // MARK: - Membership History

    func loadMembershipHistory() {
        Task {
            do { membershipHistory = try await membershipRepo.getMembershipHistory() }
            catch { self.error = error.localizedDescription }
        }
    }

    func loadMyPayments() {
        Task {
            do { myPayments = try await membershipRepo.getMyPayments() }
            catch { self.error = error.localizedDescription }
        }
    }

    func loadGallery() {
        Task {
            do { galleryPhotos = try await GalleryRepository.shared.getAll() }
            catch { /* gallery failure is non-critical */ }
        }
    }

    // MARK: - Feedback

    func loadFeedback() {
        Task {
            do { myFeedback = try await membershipRepo.getMyFeedback() }
            catch { self.error = error.localizedDescription }
        }
    }

    func submitFeedback(type: String, subject: String, description: String) {
        isLoading = true
        Task {
            do {
                let item = try await membershipRepo.submitFeedback(
                    type: type, subject: subject, description: description)
                myFeedback.insert(item, at: 0)
                successMsg = "Feedback submitted"
            } catch { self.error = error.localizedDescription }
            isLoading = false
        }
    }

    // MARK: - Contact Admin

    func loadAdminContact() {
        Task {
            do {
                adminContact = try await userRepo.getAdminContact()
            } catch { /* non-critical */ }
        }
    }

    func callAdmin() {
        guard !callAdminSent else { return }
        callAdminSent = true
        Task {
            try? await membershipRepo.callAdmin()
            try? await Task.sleep(nanoseconds: 10_000_000_000)
            callAdminSent = false
        }
    }

    func clearError() { error = nil }
    func clearSuccess() { successMsg = nil }
}
